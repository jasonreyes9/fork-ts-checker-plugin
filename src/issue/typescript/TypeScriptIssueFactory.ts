import * as ts from 'typescript'; // import for types alone
import { deduplicateAndSortIssues, Issue } from '../Issue';
import { IssueOrigin } from '../IssueOrigin';
import { IssueSeverity } from '../IssueSeverity';

function createIssueFromTsDiagnostic(
  diagnostic: ts.Diagnostic,
  typescript: typeof ts
): Issue {
  let file: string | undefined;
  let line: number | undefined;
  let character: number | undefined;

  if (diagnostic.file) {
    file = diagnostic.file.fileName;

    if (diagnostic.start) {
      const position = diagnostic.file.getLineAndCharacterOfPosition(
        diagnostic.start
      );
      line = position.line + 1;
      character = position.character + 1;
    }
  }

  return {
    origin: IssueOrigin.TYPESCRIPT,
    code: String(diagnostic.code),
    // we don't handle Suggestion and Message diagnostics
    severity:
      diagnostic.category === 0 ? IssueSeverity.WARNING : IssueSeverity.ERROR,
    message: typescript.flattenDiagnosticMessageText(
      diagnostic.messageText,
      '\n'
    ),
    file,
    line,
    character
  };
}

function createIssuesFromTsDiagnostics(
  diagnostics: ts.Diagnostic[],
  typescript: typeof ts
): Issue[] {
  function createIssueFromTsDiagnosticWithFormatter(diagnostic: ts.Diagnostic) {
    return createIssueFromTsDiagnostic(diagnostic, typescript);
  }
  return deduplicateAndSortIssues(
    diagnostics.map(createIssueFromTsDiagnosticWithFormatter)
  );
}

export { createIssueFromTsDiagnostic, createIssuesFromTsDiagnostics };
