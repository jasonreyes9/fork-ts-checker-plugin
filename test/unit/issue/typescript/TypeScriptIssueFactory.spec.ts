import {
  createIssueFromTsDiagnostic,
  createIssuesFromTsDiagnostics
} from '../../../../lib/issue';
import * as ts from 'typescript';

function makeDiagnosticMessageChainTestCase(): ts.Diagnostic {
  const version: number = Number.parseFloat(ts.versionMajorMinor);
  if (version <= 3.5) {
    return ({
      start: 12,
      code: 1221,
      category: 0,
      length: 3,
      file: undefined,
      messageText: {
        messageText: 'Cannot assign object to the string type',
        category: 0,
        code: 1221,
        next: {
          messageText: 'Another ident message',
          category: 0,
          code: 1221,
          next: {
            messageText: 'The most ident message',
            category: 0,
            code: 1221
          }
        }
      }
    } as unknown) as ts.Diagnostic;
  } else {
    // newer versions of typescript have the |next| property as an array
    return ({
      start: 12,
      code: 1221,
      category: 0,
      length: 3,
      file: undefined,
      messageText: {
        messageText: 'Cannot assign object to the string type',
        category: 0,
        code: 1221,
        next: [
          {
            messageText: 'Another ident message',
            category: 0,
            code: 1221,
            next: [
              {
                messageText: 'The most ident message',
                category: 0,
                code: 1221
              }
            ]
          }
        ]
      }
    } as unknown) as ts.Diagnostic;
  }
}

describe('[UNIT] issue/typescript/TypeScriptIssueFactory', () => {
  const TS_DIAGNOSTIC_WARNING: ts.Diagnostic = {
    start: 100,
    code: 4221,
    category: 1,
    messageText: 'Cannot assign string to the number type',
    length: 1,
    file: {
      fileName: 'src/test.ts',
      getLineAndCharacterOfPosition: () => ({
        line: 2,
        character: 2
      })
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } as any
  };
  const TS_DIAGNOSTIC_ERROR: ts.Diagnostic = {
    start: 12,
    code: 1221,
    category: 0,
    messageText: 'Cannot assign object to the string type',
    length: 1,
    file: {
      fileName: 'src/index.ts',
      getLineAndCharacterOfPosition: () => ({
        line: 5,
        character: 10
      })
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } as any
  };
  const TS_DIAGNOSTIC_WITHOUT_FILE: ts.Diagnostic = {
    start: 12,
    code: 1221,
    category: 0,
    length: 1,
    messageText: 'Cannot assign object to the string type',
    file: undefined
  };
  const TS_DIAGNOSTIC_MESSAGE_CHAIN: ts.Diagnostic = makeDiagnosticMessageChainTestCase();

  it.each([
    [TS_DIAGNOSTIC_WARNING],
    [TS_DIAGNOSTIC_ERROR],
    [TS_DIAGNOSTIC_WITHOUT_FILE],
    [TS_DIAGNOSTIC_MESSAGE_CHAIN]
  ])('creates Issue from TsDiagnostic: %p', tsDiagnostic => {
    const issue = createIssueFromTsDiagnostic(tsDiagnostic, ts);

    expect(issue).toMatchSnapshot();
  });

  it.each([
    [
      [
        TS_DIAGNOSTIC_WARNING,
        TS_DIAGNOSTIC_ERROR,
        TS_DIAGNOSTIC_WITHOUT_FILE,
        TS_DIAGNOSTIC_MESSAGE_CHAIN
      ]
    ]
  ])('creates Issues from TsDiagnostics: %p', tsDiagnostics => {
    const issues = createIssuesFromTsDiagnostics(tsDiagnostics, ts);

    expect(issues).toMatchSnapshot();
  });
});
