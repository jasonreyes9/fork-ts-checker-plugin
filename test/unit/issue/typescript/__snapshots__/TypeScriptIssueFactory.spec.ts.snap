// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[UNIT] issue/typescript/TypeScriptIssueFactory creates Issue from TsDiagnostic: {"category": 0, "code": 1221, "file": [Object], "length": 1, "messageText": "Cannot assign object to the string type", "start": 12} 1`] = `
Object {
  "character": 11,
  "code": "1221",
  "file": "src/index.ts",
  "line": 6,
  "message": "Cannot assign object to the string type",
  "origin": "typescript",
  "severity": "warning",
}
`;

exports[`[UNIT] issue/typescript/TypeScriptIssueFactory creates Issue from TsDiagnostic: {"category": 0, "code": 1221, "file": undefined, "length": 1, "messageText": "Cannot assign object to the string type", "start": 12} 1`] = `
Object {
  "character": undefined,
  "code": "1221",
  "file": undefined,
  "line": undefined,
  "message": "Cannot assign object to the string type",
  "origin": "typescript",
  "severity": "warning",
}
`;

exports[`[UNIT] issue/typescript/TypeScriptIssueFactory creates Issue from TsDiagnostic: {"category": 0, "code": 1221, "file": undefined, "length": 3, "messageText": [Object], "start": 12} 1`] = `
Object {
  "character": undefined,
  "code": "1221",
  "file": undefined,
  "line": undefined,
  "message": "Cannot assign object to the string type
  Another ident message
    The most ident message",
  "origin": "typescript",
  "severity": "warning",
}
`;

exports[`[UNIT] issue/typescript/TypeScriptIssueFactory creates Issue from TsDiagnostic: {"category": 1, "code": 4221, "file": [Object], "length": 1, "messageText": "Cannot assign string to the number type", "start": 100} 1`] = `
Object {
  "character": 3,
  "code": "4221",
  "file": "src/test.ts",
  "line": 3,
  "message": "Cannot assign string to the number type",
  "origin": "typescript",
  "severity": "error",
}
`;

exports[`[UNIT] issue/typescript/TypeScriptIssueFactory creates Issues from TsDiagnostics: [[Object], [Object], [Object], [Object]] 1`] = `
Array [
  Object {
    "character": undefined,
    "code": "1221",
    "file": undefined,
    "line": undefined,
    "message": "Cannot assign object to the string type",
    "origin": "typescript",
    "severity": "warning",
  },
  Object {
    "character": undefined,
    "code": "1221",
    "file": undefined,
    "line": undefined,
    "message": "Cannot assign object to the string type
  Another ident message
    The most ident message",
    "origin": "typescript",
    "severity": "warning",
  },
  Object {
    "character": 11,
    "code": "1221",
    "file": "src/index.ts",
    "line": 6,
    "message": "Cannot assign object to the string type",
    "origin": "typescript",
    "severity": "warning",
  },
  Object {
    "character": 3,
    "code": "4221",
    "file": "src/test.ts",
    "line": 3,
    "message": "Cannot assign string to the number type",
    "origin": "typescript",
    "severity": "error",
  },
]
`;